File: ts/utils/data/token_config.ts
import { PublicKey, Signer } from "@solana/web3.js";
import { dataStore } from "./program";
import { createRolesPDA } from ".";
import { utils } from "@coral-xyz/anchor";

// Token Config map seed.
export const TOKEN_CONFIG_MAP_SEED = utils.bytes.utf8.encode("token_config_map");

export const createTokenConfigMapPDA = (store: PublicKey) => PublicKey.findProgramAddressSync([
    TOKEN_CONFIG_MAP_SEED,
    store.toBytes(),
], dataStore.programId);

export const initializeTokenConfigMap = async (authority: Signer, store: PublicKey, len: number) => {
    const [map] = createTokenConfigMapPDA(store);
    await dataStore.methods.initializeTokenConfigMap(len).accounts({
        authority: authority.publicKey,
        store,
        onlyController: createRolesPDA(store, authority.publicKey)[0],
    }).signers([authority]).rpc();
    return map;
};

const hexStringToPublicKey = (hex: string) => {
    const decoded = utils.bytes.hex.decode(hex);
    return new PublicKey(decoded);
};

export const insertTokenConfig = async (
    authority: Signer,
    store: PublicKey,
    token: PublicKey,
    heartbeatDuration: number,
    precision: number,
    feeds: {
        pythFeedId?: string,
        chainlinkFeed?: PublicKey,
        pythDevFeed?: PublicKey,
    }
) => {
    await dataStore.methods.insertTokenConfig({
        heartbeatDuration,
        precision,
        feeds: [
            feeds.pythFeedId ? hexStringToPublicKey(feeds.pythFeedId) : PublicKey.default,
            feeds.chainlinkFeed ?? PublicKey.default,
            feeds.pythDevFeed ?? PublicKey.default,
        ]
    }, true).accounts({
        authority: authority.publicKey,
        store,
        onlyController: createRolesPDA(store, authority.publicKey)[0],
        token,
    }).signers([authority]).rpc();
};

export const insertSyntheticTokenConfig = async (
    authority: Signer,
    store: PublicKey,
    token: PublicKey,
    decimals: number,
    heartbeatDuration: number,
    precision: number,
    feeds: {
        pythFeedId?: string,
        chainlinkFeed?: PublicKey,
        pythDevFeed?: PublicKey,
    }
) => {
    await dataStore.methods.insertSyntheticTokenConfig(token, decimals, {
        heartbeatDuration,
        precision,
        feeds: [
            feeds.pythFeedId ? hexStringToPublicKey(feeds.pythFeedId) : PublicKey.default,
            feeds.chainlinkFeed ?? PublicKey.default,
            feeds.pythDevFeed ?? PublicKey.default,
        ]
    }, true).accounts({
        authority: authority.publicKey,
        store,
        onlyController: createRolesPDA(store, authority.publicKey)[0],
    }).signers([authority]).rpc();
};

export const toggleTokenConfig = async (
    authority: Signer,
    store: PublicKey,
    token: PublicKey,
    enable: boolean,
) => {
    await dataStore.methods.toggleTokenConfig(token, enable).accounts({
        authority: authority.publicKey,
        store,
        onlyController: createRolesPDA(store, authority.publicKey)[0],
    }).signers([authority]).rpc();
};

export interface TokenConfig {
    enabled: boolean,
    heartbeatDuration: number,
    tokenDecimals: number,
    precision: number,
    feeds: PublicKey[],
}

export const getTokenConfig = async (store: PublicKey, token: PublicKey) => {
    const config: TokenConfig = await dataStore.methods.getTokenConfig(store, token).accounts({
        map: createTokenConfigMapPDA(store)[0],
    }).view();
    return config;
}

export const extendTokenConfigMap = async (authority: Signer, store: PublicKey, extendLen: number) => {
    await dataStore.methods.extendTokenConfigMap(extendLen).accounts({
        authority: authority.publicKey,
        store,
        onlyController: createRolesPDA(store, authority.publicKey)[0],
    }).signers([authority]).rpc();
};

File: programs/data_store/src/instructions/token_config.rs
use anchor_lang::prelude::*;
use anchor_spl::token::Mint;

use crate::{
    states::{DataStore, Roles, Seed, TokenConfig, TokenConfigMap},
    utils::internal,
};

#[derive(Accounts)]
#[instruction(len: u16)]
pub struct InitializeTokenConfigMap<'info> {
    #[account(mut)]
    pub authority: Signer<'info>,
    pub only_controller: Account<'info, Roles>,
    pub store: Account<'info, DataStore>,
    #[account(
        init,
        space = 8 + TokenConfigMap::init_space(len as usize),
        payer = authority,
        seeds = [TokenConfigMap::SEED, store.key().as_ref()],
        bump,
    )]
    pub map: Account<'info, TokenConfigMap>,
    pub system_program: Program<'info, System>,
}

/// Initialize token config map with the given length.
pub fn initialize_token_config_map(
    ctx: Context<InitializeTokenConfigMap>,
    _len: u16,
) -> Result<()> {
    ctx.accounts.map.init(ctx.bumps.map);
    Ok(())
}

impl<'info> internal::Authentication<'info> for InitializeTokenConfigMap<'info> {
    fn authority(&self) -> &Signer<'info> {
        &self.authority
    }

    fn store(&self) -> &Account<'info, DataStore> {
        &self.store
    }

    fn roles(&self) -> &Account<'info, Roles> {
        &self.only_controller
    }
}

#[derive(Accounts)]
pub struct InsertTokenConfig<'info> {
    #[account(mut)]
    pub authority: Signer<'info>,
    pub only_controller: Account<'info, Roles>,
    pub store: Account<'info, DataStore>,
    #[account(
        mut,
        seeds = [TokenConfigMap::SEED, store.key().as_ref()],
        bump = map.bump,
        realloc = 8 + TokenConfigMap::init_space(map.length_after_insert(&token.key())),
        realloc::zero = false,
        realloc::payer = authority,
    )]
    pub map: Account<'info, TokenConfigMap>,
    pub token: Account<'info, Mint>,
    pub system_program: Program<'info, System>,
}

/// Insert or update the config of the given token.
pub fn insert_token_config(
    ctx: Context<InsertTokenConfig>,
    price_feed: Pubkey,
    heartbeat_duration: u32,
    precision: u8,
    enable: bool,
) -> Result<()> {
    let token = &ctx.accounts.token;
    ctx.accounts.map.checked_insert(
        token.key(),
        TokenConfig {
            enabled: enable,
            synthetic: false,
            price_feed,
            heartbeat_duration,
            precision,
            token_decimals: token.decimals,
        },
    )
}

impl<'info> internal::Authentication<'info> for InsertTokenConfig<'info> {
    fn authority(&self) -> &Signer<'info> {
        &self.authority
    }

    fn store(&self) -> &Account<'info, DataStore> {
        &self.store
    }

    fn roles(&self) -> &Account<'info, Roles> {
        &self.only_controller
    }
}

#[derive(Accounts)]
#[instruction(token: Pubkey)]
pub struct InsertSyntheticTokenConfig<'info> {
    #[account(mut)]
    pub authority: Signer<'info>,
    pub only_controller: Account<'info, Roles>,
    pub store: Account<'info, DataStore>,
    #[account(
        mut,
        seeds = [TokenConfigMap::SEED, store.key().as_ref()],
        bump = map.bump,
        realloc = 8 + TokenConfigMap::init_space(map.length_after_insert(&token)),
        realloc::zero = false,
        realloc::payer = authority,
    )]
    pub map: Account<'info, TokenConfigMap>,
    pub system_program: Program<'info, System>,
}

/// Insert or update the config of the given synthetic token.
pub fn insert_synthetic_token_config(
    ctx: Context<InsertSyntheticTokenConfig>,
    token: Pubkey,
    decimals: u8,
    price_feed: Pubkey,
    heartbeat_duration: u32,
    precision: u8,
    enable: bool,
) -> Result<()> {
    ctx.accounts.map.checked_insert(
        token.key(),
        TokenConfig {
            enabled: enable,
            synthetic: true,
            price_feed,
            heartbeat_duration,
            precision,
            token_decimals: decimals,
        },
    )
}

impl<'info> internal::Authentication<'info> for InsertSyntheticTokenConfig<'info> {
    fn authority(&self) -> &Signer<'info> {
        &self.authority
    }

    fn store(&self) -> &Account<'info, DataStore> {
        &self.store
    }

    fn roles(&self) -> &Account<'info, Roles> {
        &self.only_controller
    }
}

#[derive(Accounts)]
pub struct ToggleTokenConfig<'info> {
    pub authority: Signer<'info>,
    pub store: Account<'info, DataStore>,
    pub only_controller: Account<'info, Roles>,
    #[account(
        mut,
        seeds = [TokenConfigMap::SEED, store.key().as_ref()],
        bump = map.bump,
    )]
    pub map: Account<'info, TokenConfigMap>,
}

/// Toggle the config for the given token.
pub fn toggle_token_config(
    ctx: Context<ToggleTokenConfig>,
    token: Pubkey,
    enable: bool,
) -> Result<()> {
    ctx.accounts.map.toggle_token_config(&token, enable)
}

impl<'info> internal::Authentication<'info> for ToggleTokenConfig<'info> {
    fn authority(&self) -> &Signer<'info> {
        &self.authority
    }

    fn store(&self) -> &Account<'info, DataStore> {
        &self.store
    }

    fn roles(&self) -> &Account<'info, Roles> {
        &self.only_controller
    }
}

#[derive(Accounts)]
#[instruction(store: Pubkey)]
pub struct GetTokenConfig<'info> {
    #[account(
        seeds = [TokenConfigMap::SEED, store.as_ref()],
        bump = map.bump,
    )]
    pub map: Account<'info, TokenConfigMap>,
}

/// Get token config of the given token.
pub fn get_token_config(
    ctx: Context<GetTokenConfig>,
    _store: Pubkey,
    token: Pubkey,
) -> Result<Option<TokenConfig>> {
    Ok(ctx.accounts.map.as_map().get(&token).cloned())
}

#[derive(Accounts)]
#[instruction(len: u16)]
pub struct ExtendTokenConfigMap<'info> {
    #[account(mut)]
    pub authority: Signer<'info>,
    pub only_controller: Account<'info, Roles>,
    pub store: Account<'info, DataStore>,
    #[account(
        mut,
        seeds = [TokenConfigMap::SEED, store.key().as_ref()],
        bump = map.bump,
        realloc = 8 + TokenConfigMap::init_space(map.as_map().len() + len as usize),
        realloc::payer = authority,
        realloc::zero = false,
    )]
    pub map: Account<'info, TokenConfigMap>,
    pub system_program: Program<'info, System>,
}

/// Extend the length of the map with the given length.
pub fn extend_token_config_map(_ctx: Context<ExtendTokenConfigMap>, _len: u16) -> Result<()> {
    Ok(())
}

impl<'info> internal::Authentication<'info> for ExtendTokenConfigMap<'info> {
    fn authority(&self) -> &Signer<'info> {
        &self.authority
    }

    fn store(&self) -> &Account<'info, DataStore> {
        &self.store
    }

    fn roles(&self) -> &Account<'info, Roles> {
        &self.only_controller
    }
}
